File: vn/src/main/java/org/sqlbuilder/vn/objects/VnClass.java
package org.sqlbuilder.vn.objects;

import org.sqlbuilder.common.HasId;
import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.common.NotNull;
import org.sqlbuilder.common.SetCollector2;

import java.util.Comparator;
import java.util.Objects;

public class VnClass implements HasId, Insertable, Comparable<VnClass>
{
	public static final Comparator<VnClass> COMPARATOR = Comparator.comparing(VnClass::getName);

	public static final SetCollector2<VnClass> COLLECTOR = new SetCollector2<>(COMPARATOR);

	private final String name;

	public static VnClass make(final String name)
	{
		if (name == null || name.isEmpty())
		{
			throw new RuntimeException("No name");
		}

		var c = new VnClass(name);
		COLLECTOR.add(c);
		return c;
	}

	private VnClass(final String name)
	{
		this.name = name;
	}

	// A C C E S S

	public String getName()
	{
		return name;
	}

	public String getTag()
	{
		final int split = this.name.indexOf('-');
		return this.name.substring(split + 1);
	}

	@Override
	public Integer getIntId()
	{
		return COLLECTOR.apply(this);
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		VnClass that = (VnClass) o;
		return name.equals(that.name);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(name);
	}

	// O R D E R I N G

	@Override
	public int compareTo(@NotNull final VnClass that)
	{
		return COMPARATOR.compare(this, that);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return this.name;
	}

	// I N S E R T

	@Override
	public String dataRow()
	{
		// id
		// name
		// tag
		return String.format("'%s','%s'", name, getTag());
	}
}

File: vn/src/main/java/org/sqlbuilder/vn/objects/VnClass.kt
package org.sqlbuilder.vn.objects

import org.sqlbuilder.common.HasId
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.common.SetCollector
import java.util.*

class VnClass private constructor(
	@JvmField val name: String
) : HasId, Insertable, Comparable<VnClass> {

    val tag: String
        get() {
            val split = name.indexOf('-')
            return name.substring(split + 1)
        }

    override fun getIntId(): Int {
        return COLLECTOR.apply(this)
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as VnClass
        return name == that.name
    }

    override fun hashCode(): Int {
        return Objects.hash(name)
    }

    // O R D E R I N G

    override fun compareTo(that: VnClass): Int {
        return COMPARATOR.compare(this, that)
    }

    // T O S T R I N G

    override fun toString(): String {
        return name
    }

    // I N S E R T

    override fun dataRow(): String {
        return "'$name','$tag'"
    }

    companion object {

        val COMPARATOR: Comparator<VnClass> = Comparator.comparing<VnClass, String> { it.name }

        @JvmField
        val COLLECTOR: SetCollector<VnClass> = SetCollector<VnClass>(COMPARATOR)

        @JvmStatic
        fun make(name: String): VnClass {
            if (name.isEmpty()) {
                throw RuntimeException("No name")
            }

            val c = VnClass(name)
            COLLECTOR.add(c)
            return c
        }
    }
}
